### PR TITLE
First pass of #67 - Let the shim functions return an error

### PR DESCRIPTION
**Pull request by shabbyrobe**
This change is [Reviewable](https://reviewable.io/reviews/tinylib/msgp/182)

### CODE REVIEW

File: gen/decode.go [hunk 1]
```diff
@@ -146,7 +146,11 @@ func (d *decodeGen) gBase(b *BaseElem) {
 
 	// close block for 'tmp'
 	if b.Convert {
-		d.p.printf("\n%s = %s(%s)\n}", vname, b.FromBase(), tmp)
+		if b.ShimMode == Cast {
+			d.p.printf("\n%s = %s(%s)\n}", vname, b.FromBase(), tmp)
+		} else {
+			d.p.printf("\n%s, err = %s(%s)\n}", vname, b.FromBase(), tmp)
```
Comment (philhofer) [inline gen/decode.go]: Here you're generating code that shadows `err` without checking it first.

File: gen/unmarshal.go [hunk 2]
```diff
@@ -134,7 +134,12 @@ func (u *unmarshalGen) gBase(b *BaseElem) {
 	}
 	if b.Convert {
 		// close 'tmp' block
-		u.p.printf("\n%s = %s(%s)\n}", b.Varname(), b.FromBase(), refname)
+		if b.ShimMode == Cast {
+			u.p.printf("\n%s = %s(%s)\n", b.Varname(), b.FromBase(), refname)
+		} else {
+			u.p.printf("\n%s, err = %s(%s)\n", b.Varname(), b.FromBase(), refname)
```
Comment (philhofer) [inline gen/unmarshal.go]: I think you've got the same `err` shadowing problem here.

**Comment by philhofer** [comment]
Thanks for the PR.

I made a few inline comments, but overall this looks pretty good.

I wonder whether or not this should be a new directive, since `msgp:shim` is already pretty gross and verbose. Maybe `msgp:convert` or something like that.

**Comment by shabbyrobe** [comment]
Thanks heaps for the comments, can't believe I missed that error thing 😳, should be fixed now.

I started out making it a separate directive but it shared so much of the same mechanics as `//msgp:shim`. I was also thinking that it might be possible to add support for #184 via the same "shim mode" mechanism (though I haven't really thought that through fully yet).

I'm not really fussed one way or the other. If you reckon it should be a separate directive, I'll dive back in and clean it up for you, no problem.

**Comment by shabbyrobe** [comment]
Would you like me to update this to use a different directive?

**Comment by philhofer** [comment]
Sorry for the slow review time. Looks Good. Thanks again for the PR.

**Comment by shabbyrobe** [comment]
Awesome, thanks!

**Pull request merged**